### ISSUE TITLE
Array data binding in adaptive card not working

### ISSUE THREAD

**Issue by UmaPrelude** [microsoft/BotFramework-Composer#3422]
Hi,
I am trying an adaptive card with array of data mapped which i tried online which is working, but the same json returns null when i use from composer. 
Any solution? Please help.
![image](https://user-images.githubusercontent.com/65938026/84788890-47917580-b00d-11ea-83fd-28a5cfd69a59.png)
Composer
![image](https://user-images.githubusercontent.com/65938026/84790368-e79bce80-b00e-11ea-8e40-8f66cbb0a202.png)
![image](https://user-images.githubusercontent.com/65938026/84790444-fedabc00-b00e-11ea-9ee5-7fc6fef29bea.png)

**Comment by boydc2014** [comment]
Hey @UmaPrelude , the $data (data binding) feature which is a preview feature of adpativeCard called *templating*, we don't have native support that yet in Composer, in composer we used botframework LG to do data binding and send cards including adpativeCards. 

Please refer to the card samples in our samples bots to check the LG files, you can either create a sample bot or here is a direct link to the file https://github.com/microsoft/BotFramework-Composer/blob/830d27b21fa7897e1862a2736a79f2d9e71735ac/Composer/plugins/samples/assets/projects/RespondingWithCardsSample/language-generation/en-us/common.en-us.lg#L99 which then linked to another json file to hold adaptive card json. 

Basically we are doing data binding for cards using LG and AdaptiveExpression before we send to Channel, and thus not leveraging the builtin adaptiveCard data binding technique.

**Comment by hibrenda** [comment]
@UmaPrelude Composer leverage LG capability for adaptive card rendering but not leverage the adaptive card rendering strategy. @Danieladu may take a look and provide the comments.

**Comment by Danieladu** [comment]
hi @UmaPrelude. If you just want to bind the data in advance (with LG system, but not AdaptiveCards template system), below is the workaround:
````
# AdaptiveCard
[Activity
    Attachments = ${adaptivecardjson(json(employeeData()))}
]

# employeeData
-```
{
        "name": "Matt",
        "peers":
        [
            {
                "name": "Andrew",
                "title":"pm"
            }, 
            { 
                "name": "Lei",
                "title":"lead"
            }
        ]
}
```

# adaptivecardjson(employee)
- ```
{
    "$schema":"http://adaptivecards.io/schemas/adaptive-card.json",
    "type":"AdaptiveCard",
    "version":"1.2",
     "body": [
        {
            "type": "TextBlock",
            "text": "Hi ${employee.name}! Here's a bit about your org..."
        },
        {
            "type": "TextBlock",
            "text": "Your peer is: ${employee.peers[0].name} and the title is ${employee.peers[0].title}"
        }       
    ]   
}
```
````

Hope that helps. Thanks!

**Comment by UmaPrelude** [comment]
> # AdaptiveCard
> [Activity
>     Attachments = ${adaptivecardjson(json(employeeData()))}
> ]

Hi @Danieladu,

Thanks for your sugggestion.
But the thing is i am trying to dynamically  generate controls based on the number of array records.
So using the above suggestion, helps to assign the 0th index values. But what i am expecting is simply assigning the array property and generate the controls and assign its value. Which is not supporting right now. Please correct me if i am wrong.
Or can i use loop or so to achieve the same, if so please help me with the syntax.

![image](https://user-images.githubusercontent.com/65938026/84860420-9b8f6f00-b08c-11ea-8d85-b077a0634c11.png)

**Comment by UmaPrelude** [comment]
> Hey @UmaPrelude , the $data (data binding) feature which is a preview feature of adpativeCard called _templating_, we don't have native support that yet in Composer, in composer we used botframework LG to do data binding and send cards including adpativeCards.
> 
> Please refer to the card samples in our samples bots to check the LG files, you can either create a sample bot or here is a direct link to the file
> 
> https://github.com/microsoft/BotFramework-Composer/blob/830d27b21fa7897e1862a2736a79f2d9e71735ac/Composer/plugins/samples/assets/projects/RespondingWithCardsSample/language-generation/en-us/common.en-us.lg#L99
> 
> which then linked to another json file to hold adaptive card json.
> Basically we are doing data binding for cards using LG and AdaptiveExpression before we send to Channel, and thus not leveraging the builtin adaptiveCard data binding technique.

Hi @boydc2014,

Thanks for your response. I am some how trying to dynamically populate the controls in one go for number of regards and send it as a single response, so that it will send one email response on email channel. now for each record we are using one card response and it is sending separate email for each card.

Regards,
UmPrel

**Comment by Danieladu** [comment]
I get what you want....
So, The expression library provides many operations on Array/List. I'm glad to provide some solutions here:
Solution1, Embed all text in one line:
````
# AdaptiveCard
[Activity
    Attachments = ${json(adaptivecardjson(json(employeeData())))}
]

# employeeData
-```
{
        "name": "Matt",
        "peers":
        [
            {
                "name": "Andrew",
                "title":"pm"
            }, 
            { 
                "name": "Lei",
                "title":"lead"
            }
        ]
}
```

# adaptivecardjson(employee)
- ```
{
    "$schema":"http://adaptivecards.io/schemas/adaptive-card.json",
    "type":"AdaptiveCard",
    "version":"1.2",
     "body": [
        {
            "type": "TextBlock",
            "text": "Hi ${employee.name}! Here's a bit about your org..."
        },
        {
            "type": "TextBlock",
            "text": "${generateList(employee)}"
        }
    ]   
}
```

# generateList(employee)
- ${join(foreach(employee.peers, u, readout(u.name, u.title)), '\r\n')}

# readout(name, title)
- Your peer is: ${name} and the title is ${title}
````

Solution 2: Use Array operation to generate the list in json
````
# AdaptiveCard
[Activity
    Attachments = ${json(adaptivecardjson(json(employeeData())))}
]

# employeeData
-```
{
        "name": "Matt",
        "peers":
        [
            {
                "name": "Andrew",
                "title":"pm"
            }, 
            { 
                "name": "Lei",
                "title":"lead"
            }
        ]
}
```

# adaptivecardjson(employee)
- ```
{
    "$schema":"http://adaptivecards.io/schemas/adaptive-card.json",
    "type":"AdaptiveCard",
    "version":"1.2",
     "body": [
        ${generateList()}
    ]   
}
```

# generateList(employee)
- ${join(foreach(employee.peers, u, item(u.name, u.title)), ',')}

# item(name, title)
-  { "title": "${name}", "value": "${title}" }
````

Thanks!

**Comment by UmaPrelude** [comment]
Hi @Danieladu,
Thanks a lot for your support.
Solution 2 is suitable for our scenario.

Regards,
UmPrel

**Comment by Afranzio** [comment]
Hello!! I'm new to Bot Composer,
I need to pass an API response to the adaptive card and return to in table structure.

Used API: https://jsonplaceholder.typicode.com/users
Adaptive card json:
 {
    "type": "AdaptiveCard",
    "body": [
        {
            "type": "Container",
            "spacing": "Large",
            "style": "emphasis",
            "items": [
                {
                    "type": "ColumnSet",
                    "columns": [
                        {
                            "type": "Column",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "weight": "Bolder",
                                    "text": "NAME",
                                    "wrap": true
                                }
                            ],
                            "width": "auto"
                        },
                        {
                            "type": "Column",
                            "spacing": "Large",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "weight": "Bolder",
                                    "text": "EMAIL",
                                    "wrap": true,
                                    "horizontalAlignment": "Center"
                                }
                            ],
                            "width": "stretch"
                        },
                        {
                            "type": "Column",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "weight": "Bolder",
                                    "text": "WEBSITE",
                                    "wrap": true
                                }
                            ],
                            "width": "auto"
                        }
                    ]
                }
            ],
            "bleed": true
        },
        {
            "type": "Container",
            "$data": "",
            "items": [
                {
                    "type": "ColumnSet",
                    "columns": [
                        {
                            "type": "Column",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "text": "${int.name}",
                                    "wrap": true
                                }
                            ],
                            "width": "auto"
                        },
                        {
                            "type": "Column",
                            "spacing": "Medium",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "text": "${int.email}",
                                    "wrap": true,
                                    "horizontalAlignment": "Center"
                                }
                            ],
                            "width": "stretch"
                        },
                        {
                            "type": "Column",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "text": "${int.website}",
                                    "wrap": true
                                }
                            ],
                            "width": "auto"
                        }
                    ]
                }
            ]
        }
    ],
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "version": "1.2",
    "fallbackText": "This card requires Adaptive Cards v1.2 support to be rendered properly."
}

Defining function

# AdaptiveCard(res)
[Activity
    Attachments = ${json(adaptivecardjson(res))}
]


# adaptivecardjson(int)
